identity-account/src/account/account.rs

```
@@ -255,166 +252,151 @@ impl Account {
-    Ok(IdentitySnapshot {
-      sequence: commit.sequence().max(snapshot.sequence),
-      identity: commit.into_event().apply(snapshot.identity).await?,
-    })
-  }
-
-  #[doc(hidden)]
-  pub async fn load_snapshot(&self) -> Result<IdentitySnapshot> {
-    // Retrieve the state snapshot from storage or create a new one.
-    let initial: IdentitySnapshot = self
-      .storage()
-      .snapshot(self.did())
-      .await?
-      .unwrap_or_else(|| IdentitySnapshot::new(IdentityState::new()));
-
-    // Apply all recent events to the state and create a new snapshot
-    self
-      .storage()
-      .stream(self.did(), initial.sequence())
-      .await?
-      .try_fold(initial, Self::fold_snapshot)
-      .await
-  }
+  async fn publish_diff_change(&mut self, old_state: &IdentityState) -> Result<()> {
+    log::debug!("[publish_diff_change] publishing {:?}", self.document().did());
 
-  async fn load_snapshot_at(&self, generation: Generation) -> Result<IdentitySnapshot> {
-    let initial: IdentitySnapshot = IdentitySnapshot::new(IdentityState::new());
+    let old_doc: &IotaDocument = old_state.document();
+    let new_doc: &IotaDocument = self.state().document();
 
-    // Apply all events up to `generation`
-    self
-      .storage()
-      .stream(self.did(), Generation::new())
-      .await?
-      .take(generation.to_u32() as usize)
-      .try_fold(initial, Self::fold_snapshot)
-      .await
-  }
+    let mut diff_id: &MessageId = self.chain_state().last_diff_message_id();
 
-  async fn commit_events(
-    did: &IotaDID,
-    config: &AccountConfig,
-    storage: &dyn Storage,
-    state: &IdentitySnapshot,
-    events: &[Event],
-  ) -> Result<Vec<Commit>> {
-    // Bail early if there are no new events
-    if events.is_empty() {
-      return Ok(Vec::new());
+    // If there was no previous diff message, use the previous int message.
+    if diff_id.is_null() {
+      diff_id = self.chain_state.last_integration_message_id();
```

Maybe we should return an error after this if the `diff_id` is still `null`? It would mean something went wrong with the chain state but it would avoid publishing an orphaned diff update with no chain.

---

We can check, but the `diff_id` can only be null here when storing the chain state after the last update failed or the chain state got corrupted in some other way, in which case I would like to return a fatal error. When it happens, it's not the user's fault, so a fatal error seems appropriate. I would like to avoid adding another edge case error type, as that will just make life harder for the error refactoring, and instead leave a TODO there.

---

I'm not sure about the error variants but I do think we need something to indicate that the diff update will be invalid. A TODO should be fine for now.